sklearn/metrics/pairwise.py

```diff
@@ -185,12 +196,20 @@ def check_pairwise_arrays(
                 "(n_queries, n_indexed). Got (%d, %d) "
                 "for %d indexed." % (X.shape[0], X.shape[1], Y.shape[0])
             )
-    elif X.shape[1] != Y.shape[1]:
-        raise ValueError(
-            "Incompatible dimension for X and Y matrices: "
-            "X.shape[1] == %d while Y.shape[1] == %d" % (X.shape[1], Y.shape[1])
-        )
-
+    elif ensure_2d:
+        if X.shape[1] != Y.shape[1]:
+            raise ValueError(
+                "Incompatible dimension for X and Y matrices: "
+                "X.shape[1] == %d while Y.shape[1] == %d" % (X.shape[1], Y.shape[1])
+            )
+    else:
+        # the distances are neither pre-computed nor is the input expected to be 2D; we
+        # thus only check if the number of samples is the same in both arrays
+        if len(X) != len(Y):
+            raise ValueError(
+                f"Incompatible length for X and Y matrices: len(X) == {len(X)} while "
+                f" len(Y) == {len(Y)}"
+            )
```

Comment:
``check_consistent_length`` is not appropriate here since we want to ensure that both arrays have the same number of features. In fact, the added check is wrong because even when input is non-numeric, we don't want to force an equal number of samples for X and Y. I just removed it.